src/BreakpointViewer.cpp

```diff
@@ -0,0 +1,621 @@
+void BreakpointViewer::removeBreakpoint(Type type, int row)
+{
+	int index = items.indexOf({type, row});
+	// qDebug() << "removeBreakpoint, type" << type << "row" << row;
+	// for (auto item: items) {
+	// 	qDebug() << "item" << item.type << item.row;
+	// }
+	// assert(index != -1);
+
+	QString id = breakpoints->getAttribute(index, "id").toString();
+	QString cmdStr = Breakpoints::createRemoveCommand(id);
+
+	auto command = new Command(cmdStr,
+		[this] (const QString result) {
+			emit contentsUpdated(false);
+	});
+	CommClient::instance().sendCommand(command);
+}
+
+bool BreakpointViewer::processSlotField(int index, const QString& field, qint8& ps, qint8& ss)
+{
+	QStringList s = field.simplified().split("/", Qt::SplitBehaviorFlags::SkipEmptyParts);
+
+	if (s[0].compare("X", Qt::CaseInsensitive)) {
+		bool ok;
+		int value = s[0].toInt(&ok);
+		if (!ok || (value < 0 && value > 3)) {
+			// restore value
+			if (index != -1)
+				ps = breakpoints->getAttribute(index, "ps").toInt();
+			else
+				return false;
+		} else {
+			ps = value;
+		}
+	} else {
+		ps = -1;
+	}
+	if (s.length() == 2 && s[1].compare("X", Qt::CaseInsensitive)) {
+		bool ok;
+		int value = s[1].toInt(&ok);
+		if (!ok || (value < 0 && value > 3)) {
+			if (index != -1)
+				ss = breakpoints->getAttribute(index, "ss").toInt();
+			else
+				return false;
+		} else {
+			ss = value;
+		}
+	} else {
+		ss = -1;
+	}
+	return true;
+}
+
+bool BreakpointViewer::processSegmentField(int index, const QString& field, qint16& segment)
+{
+	QString s = field.simplified();
+
+	if (s.compare("X", Qt::CaseInsensitive)) {
+		bool ok;
+		int value = field.toInt(&ok);
+		if (!ok || (value < 0 && value > 255)) {
+			if (index != -1)
+				segment = breakpoints->getAttribute(index, "segment").toInt();
+			else
+				return false;
+		} else {
+			segment = value;
+		}
+	} else {
+		segment = -1;
+	}
+	return true;
+}
+
+bool BreakpointViewer::processLocationField(int index, BreakpointViewer::Type type,
+		const QString& field, int& begin, int& end)
+{
+	if (type == BreakpointViewer::BREAKPOINT) {
+		int value;
+		if ((value = stringToValue(field.simplified())) != -1) {
+			begin = 0xFFFF & value;
+			end = -1;
+			return true;
+		} else if (index != -1) {
+			begin = breakpoints->getAddress(index);
+			end = -1;
+			return true;
+		}
+		return false;
+	}
+	QStringList s = field.simplified().split(":", Qt::SplitBehaviorFlags::SkipEmptyParts);
+	int value;
+	if (!s.length() || (value = stringToValue(s[0])) == -1) {
+		if (index != -1) {
+			begin = breakpoints->getAddress(index);
+			end = breakpoints->getAttribute(index, "regionEnd").toInt();
+			return true;
+		}
+		return false;
+	}
+	begin = value;
+	if (s.length() > 2 || (s.length() == 2 && (value = stringToValue(s[1])) == -1)) {
+		if (index != -1) {
+			begin = breakpoints->getAddress(index);
+			end = breakpoints->getAttribute(index, "regionEnd").toInt();
+			return true;
+		}
+		return false;
+	}
+	end = value;
+	return true;
+}
```

Comment:
This whole code block (lines 226-246) "feels" a bit weird to me:
* Because of the duplicated `if (index != -1) { .. }` stuff.
* Because of the uninitialized `value` variable. It took me a while to verify that `value` is indeed always initialized before it's assigned to `begin` and `end`.
* Only after examining the previous point in detail, I realized that sometimes `end` is set to the same value as `begin`.
 
But I'm not exactly sure yet how to write it better. Maybe something like this (untested)?
```
    QStringList s = field.simplified().split(":", Qt::SplitBehaviorFlags::SkipEmptyParts);
    begin = (s.size() >= 1) ? stringToValue(s[0]) : -1;
    end   = (s.size() == 2) ? stringToValue(s[1]) : (s.size() == 1 ? begin : -1);
    if (begin != -1 && end != -1) return true;

    if (index == -1) return false;
    begin = breakpoints->getAddress(index);
    end = breakpoints->getAttribute(index, "regionEnd").toInt();
    return true;
```

Reply:
Testing it here it and it looks OK.

Reply:
Actually, I added the second line because openMSX will reject malformed regions with an error:
```
    end   = (s.size() == 2) ? stringToValue(s[1]) : (s.size() == 1 ? begin : -1);
    if (end < begin) return false;
    if (begin != -1 && end != -1) return true;
```